subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/logging/TaskGroupingFixture.java

```diff
@@ -0,0 +1,71 @@
+package org.gradle.integtests.fixtures.logging;
+
+import org.apache.commons.collections.map.HashedMap;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Parses rich console output into its pieces for verification in functional tests
+ */
+public class TaskGroupingFixture {
+
+    /**
+     * All tasks will start with > Task, captures everything starting with : and going until a control char
+     */
+    private final static String TASK_HEADER = "> Task (:[\\w:]*)[^\\n]*\\n";
+    private final static String BUILD_STATUS_FOOTER = "\\n[^\\n]*?BUILD SUCCESSFUL";
+    private final static String BUILD_FAILED_FOOTER = "\\n[^\\n]*?FAILURE:";
+    private final static String END_OF_TASK_OUTPUT = TASK_HEADER + "|" + BUILD_STATUS_FOOTER + "|" + BUILD_FAILED_FOOTER;
+
+    private static String PRECOMPILED_PATTERN;
+    static {
+        PRECOMPILED_PATTERN = "(?ms)";
+        PRECOMPILED_PATTERN += TASK_HEADER;
+        // Capture all output, lazily up until two new lines and an END_OF_TASK designation
+        PRECOMPILED_PATTERN += "(.*?(?=\\n\\n(?:[^\\n]*?" + END_OF_TASK_OUTPUT + ")))";
+        System.err.println(PRECOMPILED_PATTERN);
+    }
+    private static final Pattern TASK_OUTPUT_PATTERN = Pattern.compile(PRECOMPILED_PATTERN);
+
+    /**
+     * Don't need this if we parse all of the output during construction
+     */
+    private final String output;
+    private Map<String, String> taskToOutput;
+
+    public TaskGroupingFixture(String output) {
+        this.output = output;
+        parse(output);
+    }
+
+    private void parse(String output) {
+        taskToOutput = new HashedMap();
+        Matcher matcher = TASK_OUTPUT_PATTERN.matcher(output);
+        while (matcher.find()) {
+            String taskName = matcher.group(1);
+            String taskOutput = matcher.group(2);
+            if(taskToOutput.containsKey(taskName)) {
+                String combinedOutput = taskToOutput.get(taskName) + "\n" + taskOutput;
+                taskToOutput.put(taskName, combinedOutput);
+            } else {
+                taskToOutput.put(taskName, taskOutput);
+            }
+        }
+    }
+
+    public Set<String> getTaskNames() {
+        return taskToOutput.keySet();
+    }
+
+    public String getOutput(String taskName) {
+        return taskToOutput.get(taskName);
+    }
+
+    public Map<String, String> getOutputs() {
+        return taskToOutput;
+    }
+
```

Comment:
I think we want something similar to the assertion methods in this fixture:
https://github.com/gradle/gradle/blob/4b5be8babc15ad4d95b0033dda5c90fe2c1e90bd/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java#L109-L109

One of the issues with just returning a list or a map is that assertions like `outputs[':foo'] == "expected output"` look really bad when they fail (the map will be expanded with all of the keys/values, `outputs[':foo']` will be null and non-obvious).

e.g., I think we'd like to have something that looks like:

`result.groupings.task(":foo").assertOutputContains("something")`

This would fail if `:foo` could not be found.

`result.groupings.task(":foo").assertOutputEquals("something")`

This would fail if `:foo` could not be found and would automatically normalize the expected output.

`result.groupings.task(":foo").assertNoOutput()`

This would fail if `:foo` had any output.

And maybe we also want/need something like:
`result.groups.buildMessage` 

Which returns the last few lines of the build.

And 
`result.groups.ungrouped()`

which returns a List<String> of lines that weren't a part of the end of the build message or any tasks?